src/collapse/dtrace.rs

```diff
@@ -1,11 +1,16 @@
 use super::Collapse;
-use smallvec::SmallVec;
+use lazy_static::lazy_static;
+use regex::Regex;
 use std::collections::{HashMap, VecDeque};
 use std::io;
 use std::io::prelude::*;
 
 const HEADER_LINES: u8 = 3;
 
+lazy_static! {
+    static ref CPP_REGEX: Regex = Regex::new("(?P<n>::.*)[(<].*").unwrap();
```

Comment:
This is really just replicating [this regex](https://github.com/brendangregg/FlameGraph/blob/1b1c6deede9c33c5134c920bdb7a44cc5528e9a7/stackcollapse.pl#L88), right? We probably want to re-write this in straight Rust code rather than relying on a regex, as regexes are usually a fair amount slower. I'm on my way out the door, but maybe @jasonrhansen has an idea for how this can be re-written to not be a regex?

Reply:
I would use something like the following instead of using the regex in `on_stack_line`:

```Rust
        // Remove arguments from C++ function names:
        if let Some(scope) = frame.rfind("::") {
            if let Some(open) = frame[scope + 2..].find(|c| c == '(' || c == '<') {
                frame = &frame[..scope + 2 + open];
            }
        }
```

Reply:
That's not exactly the same as the regex, there are some cases where results are different on the input. But it seems like the loop is closer to 'correct'.

I'll put in some test cases on this so we can see if it makes sense to not be fully compatible

Reply:
You're correct that it doesn't match the regex exactly, but does it matter on real input?

The following lines in your tests give different results when using my code, but do lines like this ever exist in real profiles? Specifically do the places you've put `(blorg)` and `<blorg>` make sense?
```
TestClass::TestClass2(const char*)[__1cJTestClass2t6Mpkc_v_](blorg)
TestClass::TestClass2(const char*)<blorg>[__1cJTestClass2t6Mpkc_v_]
TestClass::TestClass2(const char*)[__1cJTestClass2t6Mpkc_v_]<blorg>
```

This one also gives different results:
```
TestClass::TestClass2<bla>(const char*)[__1cJTestClass2t6Mpkc_v_]
```
With regex: `TestClass::TestClass2<bla>`
With my code: `TestClass::TestClass2`

I'm wondering if not stripping the `<bla>` was intended, or a bug with the regex. If the `<bla>` isn't supposed to be stripped, then why would the regex even look for the `<` character at all?

Reply:
I generally agree the find-version seems to deliver the correcter results but I'm unsure how bug-compatible @jonhoo wants to be ;)

Reply:
I'm okay with us fixing things that are _obviously_ bugs, but I'm not so convinced by the above. But isn't it also easy to fix? Just make the `rfind` a `find` and the `find` an `rfind`? I think that should give us compatibility.

Reply:
Yeah, I'm also not sure it's a bug. That's why I asked the question. I was hoping one of you had a better I idea about what the intended behavior was. Switching `rfind` and `find` works unless there is more than one `::`.

Reply:
@jasonrhansen I think if there is more than one `::`, then
```rust
// Remove arguments from C++ function names:
if let Some(scope) = frame.find("::") {
    if let Some(open) = frame[scope + 2..].rfind(|c| c == '(' || c == '<') {
	frame = &frame[..scope + 2 + open];
    }
}
```
(that is, with the `find` and `rfind` swapped) Is still correct, no?

Reply:
Correct in the sense of matching the original regex that is :p

Reply:
Actually, I guess you're right!

Reply:
I think to find the last `::` is what we want. Given the comment, it is supposed to remove the arguments so the classpath should remain intact, not only the first class.